File: gitjuggling/src/discover.rs
use anyhow::Result;
use jwalk::WalkDir;
use std::collections::HashMap;
use std::path::{Path, PathBuf};
use std::process::Command;

/// A locally discovered git repository.
#[derive(Debug, Clone)]
pub struct LocalRepo {
    pub path: PathBuf,
    pub remote_urls: HashMap<String, String>, // remote name → URL
}

/// Discovery result with lookup maps.
#[derive(Debug)]
pub struct LocalRepos {
    pub repos: Vec<LocalRepo>,
    /// Map: remote URL (normalized) → local path
    pub by_url: HashMap<String, PathBuf>,
    /// Map: repo directory name → list of local paths
    #[allow(dead_code)]
    pub by_name: HashMap<String, Vec<PathBuf>>,
}

impl LocalRepos {
    /// Discover all git repos under the given root by scanning for `.git` directories.
    pub fn discover(root: &Path) -> Result<Self> {
        let mut repos = Vec::new();
        let mut by_url: HashMap<String, PathBuf> = HashMap::new();
        let mut by_name: HashMap<String, Vec<PathBuf>> = HashMap::new();

        let walker = WalkDir::new(root).skip_hidden(false);

        for entry in walker {
            let entry = match entry {
                Ok(e) => e,
                Err(_) => continue,
            };

            let path = entry.path();

            // Look for .git directories
            if !path.file_name().map(|n| n == ".git").unwrap_or(false) {
                continue;
            }

            // The repo path is the parent of .git/
            let repo_path = match path.parent() {
                Some(p) => p.to_path_buf(),
                None => continue,
            };

            // Read the origin remote URL
            let origin_url = get_remote_url(&repo_path, "origin");

            let mut remote_urls = HashMap::new();
            if let Some(url) = &origin_url {
                let normalized = normalize_url(url);
                by_url.insert(normalized, repo_path.clone());
                remote_urls.insert("origin".to_string(), url.clone());
            }

            // Index by directory name
            let name = repo_path
                .file_name()
                .map(|n| n.to_string_lossy().to_string())
                .unwrap_or_default();
            by_name.entry(name).or_default().push(repo_path.clone());

            repos.push(LocalRepo {
                path: repo_path,
                remote_urls,
            });
        }

        Ok(Self {
            repos,
            by_url,
            by_name,
        })
    }

    /// Find a local repo by matching against any of the provided clone URLs.
    /// Tries both the URL as-is and normalized forms.
    pub fn find_by_url(&self, url: &str) -> Option<&PathBuf> {
        let normalized = normalize_url(url);
        if let Some(p) = self.by_url.get(&normalized) {
            return Some(p);
        }
        // Also try the original URL
        if let Some(p) = self.by_url.get(url) {
            return Some(p);
        }
        None
    }
}

/// Get the URL of a git remote by name.
fn get_remote_url(repo_path: &Path, remote: &str) -> Option<String> {
    let output = Command::new("git")
        .args(["remote", "get-url", remote])
        .current_dir(repo_path)
        .output()
        .ok()?;

    if !output.status.success() {
        return None;
    }

    Some(String::from_utf8(output.stdout).ok()?.trim().to_string())
}

/// Normalize a git URL for comparison.
/// Handles SSH vs HTTPS differences for the same repo.
///
/// Examples:
///   git@github.com:owner/repo.git → github.com/owner/repo
///   https://github.com/owner/repo.git → github.com/owner/repo
///   https://git.example.com/owner/repo → git.example.com/owner/repo
fn normalize_url(url: &str) -> String {
    let url = url.trim().trim_end_matches('/');

    // SSH form: git@host:owner/repo.git
    if let Some(rest) = url.strip_prefix("git@") {
        // rest = "github.com:owner/repo.git"
        let normalized = rest.replacen(':', "/", 1);
        return normalized.trim_end_matches(".git").to_string();
    }

    // HTTPS form: https://host/owner/repo.git
    if let Some(rest) = url
        .strip_prefix("https://")
        .or_else(|| url.strip_prefix("http://"))
    {
        return rest.trim_end_matches(".git").to_string();
    }

    // Fallback: return as-is
    url.to_string()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_normalize_ssh_url() {
        assert_eq!(
            normalize_url("git@github.com:owner/repo.git"),
            "github.com/owner/repo"
        );
    }

    #[test]
    fn test_normalize_https_url() {
        assert_eq!(
            normalize_url("https://github.com/owner/repo.git"),
            "github.com/owner/repo"
        );
        assert_eq!(
            normalize_url("https://github.com/owner/repo"),
            "github.com/owner/repo"
        );
    }

    #[test]
    fn test_normalize_forgejo_url() {
        assert_eq!(
            normalize_url("https://git.example.com/user/project.git"),
            "git.example.com/user/project"
        );
    }

    #[test]
    fn test_normalize_trailing_slash() {
        assert_eq!(
            normalize_url("https://github.com/owner/repo/"),
            "github.com/owner/repo"
        );
    }

    #[test]
    fn test_normalize_idempotent() {
        let url = "github.com/owner/repo";
        assert_eq!(normalize_url(url), url);
    }
}

File: gitjuggling/internal/discover/discover.go
package discover

import (
	"fmt"
	"io/fs"
	"iter"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
)

// LocalRepo represents a locally discovered git repository.
type LocalRepo struct {
	Path       string
	RemoteURLs map[string]string // remote name → URL
	// IsWorktree marks a linked worktree of another clone. Worktrees share
	// their origin with the main clone, so they are excluded from the URL
	// index and must never be treated as an independent repo.
	IsWorktree bool
	// MainPath is the path of the main clone a worktree belongs to.
	// Only set when IsWorktree is true.
	MainPath string
}

// LocalRepos holds all discovered local repos with lookup indices.
type LocalRepos struct {
	repos     []LocalRepo
	worktrees []*LocalRepo
	byURL     map[string]string     // normalized URL → local path (main clones only)
	byPath    map[string]*LocalRepo // absolute path → LocalRepo (main clones only)
	byName    map[string][]string   // dir name → list of local paths (main clones only)
}

// NewLocalRepos builds a LocalRepos from a flat list of repos,
// constructing all lookup indices.
func NewLocalRepos(repos []LocalRepo) *LocalRepos {
	byURL := make(map[string]string, len(repos))
	byPath := make(map[string]*LocalRepo, len(repos))
	byName := make(map[string][]string)
	var worktrees []*LocalRepo

	for i := range repos {
		r := &repos[i]
		if r.IsWorktree {
			// Worktrees share their origin URL with the main clone. Indexing
			// them would let the last-discovered worktree shadow the main
			// clone in the URL lookup, so keep them out of all indexes.
			worktrees = append(worktrees, r)
			continue
		}
		for _, url := range r.RemoteURLs {
			normalized := NormalizeURL(url)
			byURL[normalized] = r.Path
		}
		byPath[r.Path] = r
		name := filepath.Base(r.Path)
		byName[name] = append(byName[name], r.Path)
	}

	return &LocalRepos{
		repos:     repos,
		worktrees: worktrees,
		byURL:     byURL,
		byPath:    byPath,
		byName:    byName,
	}
}

// Iter returns an iterator over all discovered local repos,
// including worktrees.
func (lr *LocalRepos) Iter() iter.Seq2[int, *LocalRepo] {
	return func(yield func(int, *LocalRepo) bool) {
		for i := range lr.repos {
			if !yield(i, &lr.repos[i]) {
				return
			}
		}
	}
}

// Worktrees returns the linked worktrees found during discovery.
func (lr *LocalRepos) Worktrees() []*LocalRepo {
	return lr.worktrees
}

// FindByURL looks up a local repo by matching against clone URLs.
// It tries both the normalized URL and the raw URL.
func (lr *LocalRepos) FindByURL(url string) (string, bool) {
	normalized := NormalizeURL(url)
	if p, ok := lr.byURL[normalized]; ok {
		return p, true
	}
	if p, ok := lr.byURL[url]; ok {
		return p, true
	}
	return "", false
}

// FindByPath looks up a local repo by its absolute path.
func (lr *LocalRepos) FindByPath(path string) (*LocalRepo, bool) {
	r, ok := lr.byPath[path]
	return r, ok
}

// Discover walks the filesystem under root, finding git repos by looking for
// .git entries. A .git directory is a repo (or the main clone of worktrees);
// a .git file is either a linked worktree or a submodule, told apart by
// where their gitdir points. For each entry it reads the origin remote URL.
func Discover(root string) (*LocalRepos, error) {
	var repos []LocalRepo

	err := filepath.WalkDir(root, func(path string, d fs.DirEntry, err error) error {
		if err != nil {
			return nil // skip errors, continue walking
		}

		// We only care about entries named ".git" (directory or file)
		base := filepath.Base(path)
		if base != ".git" {
			return nil
		}

		// The repo path is the parent of .git
		repoPath := filepath.Dir(path)

		var repo LocalRepo

		if d.IsDir() {
			// Regular repo or the main clone of linked worktrees.
			originURL, err := getRemoteURL(repoPath, "origin")
			if err != nil {
				// Skip repos without an origin remote (e.g. bare init, no remotes configured)
				return nil
			}
			repo = LocalRepo{
				Path:       repoPath,
				RemoteURLs: map[string]string{"origin": originURL},
			}
		} else {
			// A .git file is a linked worktree or a submodule checkout.
			mainPath, ok := worktreeMainPath(path, repoPath)
			if !ok {
				return nil // submodule or unrecognized; not a workspace repo
			}
			originURL, err := getRemoteURL(repoPath, "origin")
			if err != nil {
				return nil
			}
			repo = LocalRepo{
				Path:       repoPath,
				RemoteURLs: map[string]string{"origin": originURL},
				IsWorktree: true,
				MainPath:   mainPath,
			}
		}

		repos = append(repos, repo)

		return nil
	})
	if err != nil {
		return nil, fmt.Errorf("walking %s: %w", root, err)
	}

	return NewLocalRepos(repos), nil
}

// worktreeMainPath inspects the .git file at gitFilePath (whose parent
// directory is worktreePath) and reports the path of the main clone when the
// file points at that clone's worktree metadata (<main>/.git/worktrees/<name>).
// It returns false for anything else, notably submodule checkouts whose
// gitdir lives under <parent>/.git/modules/.
func worktreeMainPath(gitFilePath, worktreePath string) (string, bool) {
	data, err := os.ReadFile(gitFilePath)
	if err != nil {
		return "", false
	}
	content := strings.TrimSpace(string(data))
	gitdir, ok := strings.CutPrefix(content, "gitdir:")
	if !ok {
		return "", false
	}
	gitdir = strings.TrimSpace(gitdir)
	if !filepath.IsAbs(gitdir) {
		gitdir = filepath.Join(worktreePath, gitdir)
	}
	gitdir = filepath.Clean(gitdir)

	// Shape check: <main>/.git/worktrees/<name>
	nameDir := filepath.Dir(gitdir)
	if filepath.Base(nameDir) != "worktrees" {
		return "", false
	}
	dotGitDir := filepath.Dir(nameDir)
	if filepath.Base(dotGitDir) != ".git" {
		return "", false
	}
	return filepath.Dir(dotGitDir), true
}

// getRemoteURL returns the URL of a git remote by running git remote get-url.
func getRemoteURL(repoPath, remote string) (string, error) {
	cmd := exec.Command("git", "remote", "get-url", remote)
	cmd.Dir = repoPath
	out, err := cmd.Output()
	if err != nil {
		return "", err
	}
	return strings.TrimSpace(string(out)), nil
}

// NormalizeURL normalizes a git URL for comparison.
// It handles SSH vs HTTPS differences for the same repo.
//
// Examples:
//
//	git@github.com:owner/repo.git       → github.com/owner/repo
//	ssh://git@github.com/owner/repo.git → github.com/owner/repo
//	https://github.com/owner/repo.git   → github.com/owner/repo
//	https://github.com/owner/repo       → github.com/owner/repo
func NormalizeURL(url string) string {
	url = strings.TrimSpace(url)
	url = strings.TrimRight(url, "/")

	// SSH URL form: ssh://[user@]host/path.git
	if after, ok := strings.CutPrefix(url, "ssh://"); ok {
		if at := strings.Index(after, "@"); at != -1 {
			after = after[at+1:]
		}
		return strings.TrimSuffix(after, ".git")
	}

	// SCP-style SSH form: git@host:owner/repo.git
	if after, ok := strings.CutPrefix(url, "git@"); ok {
		normalized := strings.Replace(after, ":", "/", 1)
		return strings.TrimSuffix(normalized, ".git")
	}

	// HTTPS form
	if after, ok := strings.CutPrefix(url, "https://"); ok {
		return strings.TrimSuffix(after, ".git")
	}
	if after, ok := strings.CutPrefix(url, "http://"); ok {
		return strings.TrimSuffix(after, ".git")
	}

	// Fallback: return as-is
	return url
}
